### PR TITLE
8284194: Allow empty subject fields in keytool

### PR DESCRIPTION
**Pull request by wangweij**
This code change allows one entering "." at a distinguished name prompt to skip a sub-component when running `keytool -genkeyapir`. Several new resource strings are added.

There is no detailed description in `keytool.html`, so I think there's no need to update it.

I'll file a CSR to describe the behavior change.

Here is an example after this change:
```
$ keytool -genkeypair -keystore ks -storepass changeit -alias b -keyalg EC
Enter the distinguished name. Provide a single dot (.) to leave a sub-component empty or press ENTER to use the default value in braces.
What is your first and last name?
  [Unknown]:  .
What is the name of your organizational unit?
  [Unknown]:  .
What is the name of your organization?
  [Unknown]:  .
What is the name of your City or Locality?
  [Unknown]:  .
What is the name of your State or Province?
  [Unknown]:  .
What is the two-letter country code for this unit?
  [Unknown]:  .
At least one field must be provided. Enter again.
Enter the distinguished name. Provide a single dot (.) to leave a sub-component empty or press ENTER to use the default value in braces.
What is your first and last name?
  []:  Duke
What is the name of your organizational unit?
  []:
What is the name of your organization?
  []:
What is the name of your City or Locality?
  []:
What is the name of your State or Province?
  []:
What is the two-letter country code for this unit?
  []:
Is CN=Duke correct?
  [no]:  yes

Generating 384 bit EC (secp384r1) key pair and self-signed certificate (SHA384withECDSA) with a validity of 90 days
	for: CN=Duke
```
In the first round, "." is entered for all fields and keytool rejected it. In the second round, CN is entered but the others are unchanged (just type enter, because they are already entered previously). At the end, the name is "CN=Duke".

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved
- [x] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issues
 * [JDK-8284194](https://bugs.openjdk.java.net/browse/JDK-8284194): Allow empty subject fields in keytool
 * [JDK-8286614](https://bugs.openjdk.java.net/browse/JDK-8286614): Allow empty subject fields in keytool (**CSR**)

### Reviewers
 * [Jamil Nimeh](https://openjdk.java.net/census#jnimeh) (@jnimeh - **Reviewer**) ⚠️ Review applies to [abed47cb](https://git.openjdk.java.net/jdk/pull/8667/files/abed47cb4ab3aab39133babcdef382a3c6177aaa)
 * [Hai-May Chao](https://openjdk.java.net/census#hchao) (@haimaychao - Committer) ⚠️ Review applies to [abed47cb](https://git.openjdk.java.net/jdk/pull/8667/files/abed47cb4ab3aab39133babcdef382a3c6177aaa)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8667/head:pull/8667` \
`$ git checkout pull/8667`

Update a local copy of the PR: \
`$ git checkout pull/8667` \
`$ git pull https://git.openjdk.java.net/jdk pull/8667/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8667`

View PR using the GUI difftool: \
`$ git pr show -t 8667`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8667.diff">https://git.openjdk.java.net/jdk/pull/8667.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/sun/security/tools/keytool/Main.java [hunk 1]
```diff
@@ -3756,28 +3755,40 @@ private X500Name getX500Name() throws IOException {
...
+                        collator.compare(userInput, rb.getString("y")) != 0;
+            }
+        } while (needRepeat);
 
         System.err.println();
         return name;
     }
 
+    private static String dotToNull(String input) {
+        return ".".equals(input) ? null : input;
```
Comment (jnimeh) [inline src/java.base/share/classes/sun/security/tools/keytool/Main.java:3781]: Do we deal at all with leading/trailing whitespace (maybe more concerned about trailing whitespace than leading)?  What happens if we get a ". " (trailing space)?

**Review by jnimeh — COMMENTED** [review]
Looks pretty solid.  Just that one question about how to handle an errant space after the ".".  I think it might be fine to treat it literally as a dot-space combo.  If it was an error it would be caught by the user most likely at the check at the end of the DN entry loop.

File: src/java.base/share/classes/sun/security/tools/keytool/Main.java (hunk 1 above)
Comment (wangweij) [reply src/java.base/share/classes/sun/security/tools/keytool/Main.java:3781]: The user must be deliberately doing this. Let's respect their decision. :-)

**Review by jnimeh — APPROVED** [review]

**Review by haimaychao — APPROVED** [review]

**Comment by seanjmullan** [comment]
It might also be helpful to note that hitting return or enter will use the default, ex: "Enter a single dot (.) to leave the sub-component empty or return to use the default value."

For this:

> What is your first and last name?
>   [EMPTY]:  Duke

I find the word "EMPTY" here a bit confusing because this is not a default value like "Unknown". It seems to me that it might be more intuitive to just repeat the initial set of prompts using [Unknown] and requiring '.' to be entered, especially since you repeat the part "Enter a single dot (.) to leave the sub-component empty."

**Comment by wangweij** [comment]
I've already modified the prompt a little before the CSR is finalized. How about
```
Provider a single dot (.) to leave a sub-component empty or press ENTER to use the default value in braces.
```
Hopefully macOS users know "ENTER" is "return" and people won't debate on "braces" or "brackets".

As for the "[EMPTY]" prompt, I think people have already entered "." in the first round and we always remember their inputs as the new default value in the next round so that they only need to enter the components they want to update.

**Comment by seanjmullan** [comment]
> I've already modified the prompt a little before the CSR is finalized. How about
> 
> ```
> Enter the distinguished name. Provider a single dot (.) to leave a sub-component empty or press ENTER to use the default value in braces.
> ```
> 
> Hopefully this is not too long, and macOS users know "ENTER" is "return", and people won't debate on "braces" or "brackets".

Ok, I like that. One typo though: s/Provider/Provide

> As for the "[EMPTY]" prompt, the user has already entered "." in the first round and we always remember the inputs as the new default values in the next round so that they only need to enter the components they want to update.

What about just "[]"? I think if you keep EMPTY, you should really define what that keyword means, but that might make the text too verbose.

**Pull request closed without merging**